File: apps/android/core/ui/src/main/kotlin/app/gauja/core/ui/state/ContentMessage.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.ui.state

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.tooling.preview.PreviewScreenSizes
import androidx.compose.ui.unit.dp
import app.gauja.core.common.AppError
import app.gauja.core.designsystem.GaujaTheme
import app.gauja.core.ui.R

@Composable
fun ContentMessage(error: AppError, modifier: Modifier = Modifier, onRetry: (() -> Unit)? = null) {
    val message =
        when (error) {
            AppError.NETWORK -> R.string.error_network
            AppError.TLS -> R.string.error_tls
            AppError.AUTH -> R.string.error_auth
            AppError.PERMISSION -> R.string.error_permission
            AppError.NOT_FOUND -> R.string.error_notfound
            AppError.VALIDATION -> R.string.error_validation
            AppError.SERVER_VERSION -> R.string.error_serverversion
            AppError.OFFLINE -> R.string.error_offline
            AppError.UNKNOWN -> R.string.error_unknown
        }
    Column(modifier.padding(16.dp).semantics { liveRegion = LiveRegionMode.Polite }) {
        Text(stringResource(message))
        if (onRetry != null) Button(onClick = onRetry) { Text(stringResource(R.string.retry)) }
    }
}

@Preview
@PreviewScreenSizes
@Composable
private fun ErrorPreview() {
    GaujaTheme { ContentMessage(AppError.OFFLINE) }
}

File: apps/ios/Packages/UI/Sources/UI/state/ContentMessage.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Common
import DesignSystem
import SwiftUI

public struct ContentMessage: View {
    private let error: AppError
    private let retry: (() -> Void)?
    public init(error: AppError, retry: (() -> Void)? = nil) {
        self.error = error
        self.retry = retry
    }
    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(message)
            if let retry { Button("Retry", action: retry).buttonStyle(.borderedProminent) }
        }.padding().accessibilityElement(children: .contain)
    }
    private var message: LocalizedStringKey {
        switch error {
        case .network: "The server could not be reached. Try again."
        case .tls: "The secure connection could not be verified. Review this server’s trust settings."
        case .auth: "Sign in to this server again."
        case .permission: "You do not have permission to view this content."
        case .notFound: "This content was not found."
        case .validation: "The server response or input was invalid."
        case .serverVersion: "This feature is not supported by this server version."
        case .offline: "You are offline. Cached content remains available."
        case .unknown: "Saved data could not be read. Try reopening Gauja."
        }
    }
}
#Preview("Offline") { GaujaTheme { ContentMessage(error: .offline) } }
#Preview("Largest text", traits: .fixedLayout(width: 500, height: 600)) {
    GaujaTheme { ContentMessage(error: .network, retry: {}).environment(\.dynamicTypeSize, .accessibility5) }
}
